Cover image: added upload button.

jorgefilipecosta opened the pull request:
This PR adds an upload button to cover image to make it consistent with the image block.
Fixes: https://github.com/WordPress/gutenberg/issues/4341


## How Has This Been Tested?
Add a cover image, use the upload button select an image and verify the image was used in the cover image block.

## Screenshots (jpeg or gifs if applicable):
<img width="633" alt="screen shot 2018-01-08 at 23 01 15" src="https://user-images.githubusercontent.com/11271197/34696813-12bdf958-f4c8-11e7-8b21-9340eec223f1.png">

jorgefilipecosta committed:
Cover image block: Refactored to use the new image placeholder helper.

For consistency reasons with image block.

mtias commented:
It would be great if all these became an `ImagePlaceholder` component that came with the upload buttons, drag area, etc. The text and title should probably be customizable.

jorgefilipecosta committed:
Implemented image placeholder block helper.

jorgefilipecosta committed:
Image block: Refactored to use the new image placeholder helper.

youknowriad commented on blocks/image-placeholder/index.js:
```diff
@@ -0,0 +1,42 @@
+ */
+import MediaUploadButton from '../media-upload-button';
+
+export default function ImagePlaceHolder( { className, setAttributes, icon, label, onSelectImage } ) {
```
Great work extracting this, there's something that bother me a bit about the `setAttributes` being passed here and also being passed to `mediaUpload` (probably the root issue).

Instead of passing `setAttributes` to `mediaUpload`, should we refactor it to pass something like `onChange` and `onChange` would be called with a single or multiple images (it's used in gallery aswell).

This would allow us to only pass `onChange` here.

The issue with the current approach is that this "reusable" component enforces the block using it to have certain attribute keys.

jorgefilipecosta replied:
I totally agree with this changes, the refactor should be applied in the media upload because it is there that we are using the attribute keys. I will add to this PR a refactor to the media upload as previous work.

jorgefilipecosta replied:
A refactor was applied to the mediaUpload. It is now more generic and does not contain block specific logic e.g: for the gallery.
The placeholder now does not receive setAttributes it just receives onSelectImage that is used for both uploaded and gallery images. An object is passed to function describing the image, the user of the component can pass a function that maps the image object to its attributes.

jorgefilipecosta committed:
Refactored mediaUpload to be more generic.

mediaUpload now receives receives an array of files and a function that it calls with an array of objects representing the files uploaded. Logic specific to gallery and image blocks was removed from the function.

youknowriad reviewed (approved):
This needs a rebase, but awesome work on this PR
